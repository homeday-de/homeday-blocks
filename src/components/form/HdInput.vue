<template>
  <div
    :class="fieldClasses"
    class="field field--input"
  >
    <img
      v-if="icon"
      :src="icon"
      role="presentation"
      class="field__icon"
    >
    <input
      :autocomplete="autocomplete"
      :value="value"
      :type="currentType"
      :id="name"
      :name="name"
      :placeholder="isActive && placeholder !== undefined ? placeholder : ''"
      :required="required"
      :autofocus="autofocus"
      :disabled="disabled"
      class="field__input"
      ref="input"
      data-lpignore="true"
      @input="handleInput"
      @focus="handleFocus"
      @blur="handleBlur"
    />
    <label
      v-if="label"
      :for="name"
      class="field__label"
    >
      {{ label }}
    </label>
    <p
      v-if="error"
      class="field__error"
    >
      {{ error }}
    </p>
    <p
      v-else-if="helper"
      class="field__error field__error--helper"
      v-html="helper"
    />
    <span class="field__border"/>
  </div>
</template>

<script>
import merge from 'lodash/merge';
import { getMessages } from 'homeday-blocks/src/lang';
import {
  email as validateEmail,
  date as validateDate,
} from 'homeday-blocks/src/services/formValidation';

export default {
  name: 'HdInput',
  props: {
    label: {
      type: String,
      default: '',
    },
    name: {
      type: String,
      required: true,
    },
    type: {
      type: String,
      default: 'text',
    },
    value: {
      type: [String, Number],
      default: '',
    },
    placeholder: {
      type: String,
      default: '',
    },
    required: {
      type: Boolean,
      default: false,
    },
    autocomplete: {
      type: String,
      default: 'on',
    },
    autofocus: {
      type: Boolean,
      default: false,
    },
    min: {
      type: Number,
      default: undefined,
    },
    max: {
      type: Number,
      default: undefined,
    },
    lang: {
      type: String,
      default: 'de',
    },
    texts: {
      type: Object,
      default: () => ({}),
    },
    disabled: {
      type: Boolean,
      default: false,
    },
    icon: {
      type: String,
      default: '',
    },
    customRules: {
      type: Array,
      default: () => [],
      validator: rulesProvided => rulesProvided.every(
        ({ validate, errorMessage }) => typeof validate === 'function' && typeof errorMessage === 'string',
      ),
    },
  },
  data() {
    return {
      currentType: this.type || 'text',
      isActive: undefined,
      isValid: undefined,
      error: null,
      helper: null,
    };
  },
  computed: {
    t() {
      return merge(getMessages(this.lang), this.texts);
    },
    isEmpty() {
      return this.value === null || this.value === undefined || this.value === '';
    },
    fieldClasses() {
      return {
        'field--active': this.isActive,
        'field--filled': !this.isEmpty,
        'field--invalid': this.isValid === false,
        'field--disabled': this.disabled,
        'field--hasIcon': this.icon,
      };
    },
  },
  watch: {
    value() {
      this.validate();
    },
    type(type) {
      this.currentType = type;
    },
  },
  methods: {
    clearInput() {
      this.$refs.input.focus();
      this.hideError();
      this.$emit('input', '');
    },
    handleFocus() {
      this.isActive = true;
      this.$emit('focus');
    },
    handleBlur() {
      this.isActive = false;
      this.validate();
      this.$emit('blur');
    },
    handleInput(e) {
      let newValue = e.target.value;

      if (this.currentType === 'number' && newValue !== '') {
        newValue = parseFloat(newValue);

        if (typeof this.min === 'number' && newValue < this.min) {
          newValue = this.min;
        } else if (typeof this.max === 'number' && newValue > this.max) {
          newValue = this.max;
        }

        if (newValue === this.value) {
          this.$nextTick(this.$forceUpdate);
        }
      }

      this.$emit('input', newValue);
    },
    showError(errorMessage) {
      this.error = errorMessage;
      this.isValid = false;
    },
    showHelper(message) {
      this.helper = message;
    },
    hideError() {
      this.isValid = true;
      this.error = null;
    },
    validate() {
      if (this.required && this.isEmpty) {
        this.showError(this.t.FORM.VALIDATION.REQUIRED);
      } else if (this.currentType === 'email' && !validateEmail(this.value)) {
        this.showError(this.t.FORM.VALIDATION.INVALID_EMAIL);
      } else if (this.currentType === 'date' && !validateDate(this.value)) {
        this.showError(this.t.FORM.VALIDATION.INVALID_DATE);
      } else if (this.customRules.length && !this.isEmpty) {
        this.validateCustomRules();
      } else {
        this.hideError();
      }

      return !this.error;
    },
    validateCustomRules() {
      const firstFailingRule = this.customRules.find(({ validate }) => !validate(this.value));
      if (firstFailingRule) {
        this.showError(firstFailingRule.errorMessage);
      } else {
        this.hideError();
      }
    },
  },
};
</script>

<!-- Add "scoped" attribute to limit CSS to this component only -->
<style scoped lang="scss">
@import 'homeday-blocks/src/styles/mixins.scss';
@import 'homeday-blocks/src/styles/inputs.scss';

.field {
  &__error {
    width: 100%;
    text-align: left;
    &--helper {
      display: block;
      color: getShade($quaternary-color, 80);
    }
  }
}
</style>
