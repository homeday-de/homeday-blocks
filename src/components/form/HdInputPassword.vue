<template>
  <div class="password-input">
    <HdInput
      ref="input"
      :type="type"
      v-model="computedValue"
      v-bind="$attrs"
      :class="{ 'field--hasControl': this.showVisibilityToggle }"
    />

    <span v-if="showVisibilityToggle"
      class="field__visibilityToggle"
      :class="{'field__visibilityToggle--visible': type === 'text'}"
      @click="toggleVisibility"
    >
      <HdIcon
        :src="visibilityOnIcon"
        width="100%"
        height="100%"
      />
    </span>
  </div>
</template>

<script>
import HdInput from 'homeday-blocks/src/components/form/HdInput.vue';
import HdIcon from 'homeday-blocks/src/components/HdIcon.vue';
import { visibilityOnIcon } from 'homeday-blocks/src/assets/small-icons';

export default {
  name: 'HdInputPassword',
  inheritAttrs: false,
  components: {
    HdInput,
    HdIcon,
  },
  props: {
    value: {
      type: String,
      default: '',
    },
  },
  data() {
    return {
      type: 'password',
      visibilityOnIcon,
    };
  },
  computed: {
    computedValue: {
      get() { return this.value; },
      set(value) {
        this.$emit('input', value);
      },
    },
    showVisibilityToggle() {
      return this.value.length !== 0;
    },
  },
  methods: {
    toggleVisibility() {
      this.type = this.type === 'password' ? 'text' : 'password';
    },
    showError(...args) {
      return this.$refs.input.showError(...args);
    },
    showHelper(...args) {
      return this.$refs.input.showHelper(...args);
    },
    validate(...args) {
      return this.$refs.input.validate(...args);
    },
    hideError(...args) {
      return this.$refs.input.hideError(...args);
    },
  },
};
</script>

<!-- Add "scoped" attribute to limit CSS to this component only -->
<style scoped lang="scss">
@import 'homeday-blocks/src/styles/mixins.scss';
@import 'homeday-blocks/src/styles/inputs.scss';

$icon-color: getShade($quaternary-color, 80);

.passwordInput {
  position: relative;

  .field__visibilityToggle {

    position: absolute;
    top: $inputPaddingTop;
    right: $inline-s;
    width: 24px;
    height: 24px;
    cursor: pointer;
    &:after {
      display: block;
      content: '';
      position: absolute;
      opacity: 1;
      top: 2px;
      left: 3px;
      width: calc(100% - 1px);
      height: 0px;
      border-bottom: 2px solid $icon-color;
      border-top: 1px solid $inputBackground;
      transform-origin: left;
      transform: rotateZ(45deg) scaleX(1);
      transition: transform .3s;
    }
    &--visible {
      &:after {
        transform: rotateZ(45deg) scaleX(0);
      }
    }

    &::v-deep path {
      fill: $icon-color;
    }
  }
}
</style>
