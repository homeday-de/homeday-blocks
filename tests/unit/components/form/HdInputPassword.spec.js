import { wrapperFactoryBuilder } from 'tests/unit/helpers';
import HdInputPassword from '@/components/form/HdInputPassword.vue';
import HdInput from '@/components/form/HdInput.vue';

const VISIBILITY_TOGGLE_SELECTOR = '.field__visibilityToggle';

const PASSWORD_VISIBLE_CLASS = 'field__visibilityToggle--visible';

const TEST_VALUE = 'new value';
const TEST_PROPS = {
  name: 'test name',
  label: 'test label',
  placeholder: 'test placeholder',
  value: 'test password',
};

describe('HdInputPassword', () => {
  const wrapperFactory = wrapperFactoryBuilder(
    HdInputPassword,
    {
      propsData: TEST_PROPS,
    },
  );

  let wrapper;
  let inputWrapper;
  beforeEach(() => {
    wrapper = wrapperFactory();
    inputWrapper = wrapper.find(HdInput);
  });

  test('is rendered as expected', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('updates the field on change of the prop `value`', () => {
    wrapper.setProps({ value: TEST_VALUE });
    expect(wrapper.find('input').element.value).toBe(TEST_VALUE);
  });

  test('emits `input` event with the right payload', () => {
    wrapper.find('input').setValue(TEST_VALUE);
    expect(inputWrapper.emitted('input')[0][0]).toEqual(TEST_VALUE);
  });

  test('check if the input type is "password"', () => {
    expect(wrapper
      .find('input')
      .attributes()
      .type).toBe('password');
  });

  test('shows the visibility toggle', () => {
    expect(wrapper.find(VISIBILITY_TOGGLE_SELECTOR).exists()).toBe(true);
  });

  test('toggles the password\'s visibility', () => {
    const toggle = wrapper.find(VISIBILITY_TOGGLE_SELECTOR);

    toggle.trigger('click');
    expect(wrapper.find('input').attributes().type).toBe('text');
    expect(toggle.classes()).toContain(PASSWORD_VISIBLE_CLASS);

    toggle.trigger('click');
    expect(wrapper.find('input').attributes().type).toBe('password');
    expect(toggle.classes()).not.toContain(PASSWORD_VISIBLE_CLASS);
  });

  test('doesn\'t show visibility toggle if empty', () => {
    wrapper.setProps({ value: '' });
    expect(wrapper.find(VISIBILITY_TOGGLE_SELECTOR).exists()).toBe(false);
  });
});
