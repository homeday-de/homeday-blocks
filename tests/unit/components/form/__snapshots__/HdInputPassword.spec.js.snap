// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HdInputPassword is rendered as expected 1`] = `
<div class="password-input">
  <div class="field field--input field--filled field--hasControl">
    <!----> <input autocomplete="on" type="password" id="test name" name="test name" placeholder="" data-lpignore="true" class="field__input"> <label for="test name" class="field__label">
      test label
    </label>
    <!----> <span class="field__border"></span></div> <span class="field__visibilityToggle"><!----></span>
</div>
`;
